### PR TITLE
Implemented caching with Azure containers using sccache

### PR DESCRIPTION
**Pull request by Eliasj42**
ci-exactly: linux_x64_clang

### CODE REVIEW

File: .github/workflows/ci_linux_x64_clang.yml [hunk 1]
```diff
@@ -42,8 +45,14 @@ jobs:
...
         run: python3 -m pip install -r ./runtime/bindings/python/iree/runtime/build_requirements.txt
+      - name: Setting up Cache
+        run: |
+          mkdir -p /mnt/azureblob
+          python3 ./build_tools/ccache/edit_fuse_connection.py ${{ secrets.AZURE_CCACHE_CONTAINER_KEY }}
+          blobfuse2 mount --allow-other --config-file=./build_tools/ccache/fuse_connection2.yaml /mnt/azureblob/
+
       # TODO(#18238): add local or remote ccache
-      - name: CMake - configure
+      - name: CMake - configurea
```
Comment (Eliasj42) [inline .github/workflows/ci_linux_x64_clang.yml]: Oops, don't know how that got in there, fixing it after I make sure tests pass

File: .github/workflows/ci_linux_x64_clang.yml [hunk 2]
```diff
@@ -29,7 +29,10 @@ jobs:
     needs: setup
     if: contains(fromJson(needs.setup.outputs.enabled-jobs), 'linux_x64_clang')
     runs-on: azure-linux-scale
-    container: ghcr.io/iree-org/cpubuilder_ubuntu_jammy_x86_64@sha256:ba10d33ef502d8b33aaa43800e640a6866983ef353b73fda2ed6595c4275e066
+    container:
+      # blobfuse requires the privileged flag
+      image: ghcr.io/iree-org/cpubuilder_ubuntu_jammy_x86_64:main
```
Comment (Eliasj42) [inline .github/workflows/ci_linux_x64_clang.yml]: Thoughts on this? using the hash I couldn't update the container, but do we always want to use the most up to date

Comment (ScottTodd) [reply .github/workflows/ci_linux_x64_clang.yml]: My concern with using `:main` is that if we push a bad update, workflows will be immediately affected by it. I'd rather pin to a version and then choose to rollout the update intentionally (with checks running on the update).

Comment (Eliasj42) [reply .github/workflows/ci_linux_x64_clang.yml]: ok ill find a hash of the new version

**Comment by saienduri** [comment]
image

Do some of these have to be enabled?

File: .github/workflows/ci_linux_x64_clang.yml [hunk 3]
```diff
@@ -54,6 +63,7 @@ jobs:
             -DIREE_ENABLE_LLD=ON \
             -DIREE_ENABLE_ASSERTIONS=ON \
             -DIREE_BUILD_DOCS=ON \
+            -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
             -DIREE_TARGET_BACKEND_WEBGPU_SPIRV=ON
       - name: CMake - build
```
Comment (ScottTodd) [inline .github/workflows/ci_linux_x64_clang.yml:65]: Can you add a `ccache --show-stats` call after the build step, then push a few commits? I'd expect the first run to populate the cache and subsequent runs to use it.

Something like this:

```diff
 - name: CMake - build
        run: |
+         ccache --zero-stats
+         ccache --show-stats
          cmake --build ${BUILD_DIR} -- -k 0
          cmake --build ${BUILD_DIR} --target install -- -k 0
          cmake --build ${BUILD_DIR} --target iree-test-deps -- -k 0
+         ccache --show-stats
```

**Comment by ScottTodd** [comment]
> image
>
> Do some of these have to be enabled?

I suggested that Elias start with `linux_x64_clang`. That is the simplest workflow and also one the main ones. I can help get other workflows and scripts updated once we have one working.

**Comment by saienduri** [comment]
Looks like latest run is still not really utilizing cache:

image

Do we just need another run or something else?

**Comment by ScottTodd** [comment]
> Looks like latest run is still not really utilizing cache:
> image
> 
> Do we just need another run or something else?

Something else... probably.

We've been chatting / debugging in the Slack thread all day :P. I just posted an update here: https://github.com/iree-org/iree/issues/18238#issuecomment-2339319989

**Comment by saienduri** [comment]
Ah I see, thanks for the update

File: .github/workflows/ci_linux_x64_clang.yml [hunk 4]
```diff
@@ -42,9 +45,19 @@ jobs:
           submodules: true
       - name: Install Python requirements
         run: python3 -m pip install -r ./runtime/bindings/python/iree/runtime/build_requirements.txt
+      - name: Setting up Cache
+        run: |
+          mkdir -p /mnt/azureblob
+          python3 ./build_tools/ccache/edit_fuse_connection.py ${{ secrets.AZURE_CCACHE_CONTAINER_KEY }}
+          blobfuse2 mount --allow-other --config-file=./build_tools/ccache/fuse_connection2.yaml /mnt/azureblob/
+          export SCCACHE_DIR=/mnt/azureblob/ccache-container
```
Comment (ScottTodd) [inline .github/workflows/ci_linux_x64_clang.yml]: Shouldn't need this mount or any of the `build_tools/ccache/` files when using sccache, unless we use the mount for anonymous read access. I'd start with the authenticated read/write access.

File: .github/workflows/ci_linux_x64_clang.yml [hunk 5]
```diff
@@ -42,9 +45,19 @@ jobs:
...
+
       # TODO(#18238): add local or remote ccache
       - name: CMake - configure
         run: |
+          export SCCACHE_AZURE_CONNECTION_STRING=${{ secrets.SCCACHE_AZURE_CONNECTION_STRING  }}
+          export SCCACHE_AZURE_BLOB_CONTAINER="ccache-container"
```
Comment (ScottTodd) [inline .github/workflows/ci_linux_x64_clang.yml]: You can set `SCCACHE_AZURE_KEY_PREFIX` here to something matching the job name (e.g. `linux_x64_clang` here). I'll write up more thoughts about namespacing on the infra tracking issue.

Comment (ScottTodd) [reply .github/workflows/ci_linux_x64_clang.yml]: https://github.com/iree-org/iree/issues/18238#issuecomment-2341505960

File: .github/workflows/ci_linux_x64_clang.yml [hunk 6]
```diff
@@ -29,12 +29,17 @@ jobs:
...
     runs-on: azure-linux-scale
-    container: ghcr.io/iree-org/cpubuilder_ubuntu_jammy_x86_64@sha256:ba10d33ef502d8b33aaa43800e640a6866983ef353b73fda2ed6595c4275e066
+    container:
+      image: ghcr.io/iree-org/cpubuilder_ubuntu_jammy_x86_64@sha256:2b2ad51d7de988be13086bc618d89d2ba47fbf09eb5b38c60dce82b595fb1c74
     defaults:
       run:
         shell: bash
     env:
       BUILD_DIR: build
+      SCCACHE_AZURE_CONNECTION_STRING: "${{ secrets.AZURE_CCACHE_CONNECTION_STRING }}"
```
Comment (ScottTodd) [inline .github/workflows/ci_linux_x64_clang.yml:39]: We're having some offline discussions about readonly vs readwrite and PRs from the main repo and from forks.

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**